File: behavioral/typescript/iterator_structured/tree_iterator.ts
import { Iterator } from './interfaces';

export class TreeNode<T> {
  data: T;
  left: TreeNode<T> | null = null;
  right: TreeNode<T> | null = null;
  
  constructor(data: T) {
    this.data = data;
  }
}

export class BinaryTreeIterator<T> implements Iterator<T> {
  private stack: TreeNode<T>[] = [];
  
  constructor(root: TreeNode<T> | null) {
    this.pushLeft(root);
  }
  
  private pushLeft(node: TreeNode<T> | null): void {
    while (node) {
      this.stack.push(node);
      node = node.left;
    }
  }
  
  hasNext(): boolean {
    return this.stack.length > 0;
  }
  
  next(): T {
    if (!this.hasNext()) return null as T;
    
    const node = this.stack.pop()!;
    
    if (node.right) {
      this.pushLeft(node.right);
    }
    
    return node.data;
  }
}

File: behavioral/dart/iterator_structured/tree_iterator.dart
import 'interfaces.dart';

class TreeNode<T> {
  T data;
  TreeNode<T>? left;
  TreeNode<T>? right;
  
  TreeNode(this.data);
}

class BinaryTreeIterator<T> implements Iterator<T> {
  final List<TreeNode<T>> _stack = [];
  
  BinaryTreeIterator(TreeNode<T>? root) {
    _pushLeft(root);
  }
  
  void _pushLeft(TreeNode<T>? node) {
    while (node != null) {
      _stack.add(node);
      node = node.left;
    }
  }
  
  @override
  bool hasNext() {
    return _stack.isNotEmpty;
  }
  
  @override
  T next() {
    if (!hasNext()) return null as T;
    
    final node = _stack.removeLast();
    
    if (node.right != null) {
      _pushLeft(node.right);
    }
    
    return node.data;
  }
}
